libc/stdio/stdio_indicator.c

```diff
@@ -0,0 +1,628 @@
+		TEST_ASSERT_NOT_NULL_MESSAGE(f, tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, feof(f), tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL(-1, fgetc(f));
+
+		TEST_ASSERT_EQUAL_MESSAGE(CHECK_MATCH_FIRST_STRING(modes[i], "w", "a", "wb", "ab") ? 0 : 1, feof(f), tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, fclose(f), tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, feof(f), tellMode(modes[i]));
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, feof(f), tellMode(modes[i]));
+	}
+	TEST_ASSERT_EQUAL(0, remove(filename));
+}
+
+
+TEST(stdio_feof, preserve_errno_huge_size)
+{
+	int errnoBefore;
+	int multiplier = 10;
+	long pos = 0;
+	const char *filename = "test_stdio_feof_errno_preserve";
+	FILE *f = fopen(filename, "w+");
+	TEST_ASSERT_NOT_NULL(f);
+
+	for (int i = 0; i < multiplier; i++) {
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fputs(testdata_hugeStr, f));
+	}
+
+	fputc('A', f);
+	errno = 0;
+	while (pos <= multiplier * testdata_hugeSize) {
+		TEST_ASSERT_EQUAL_INT(0, fseek(f, pos, SEEK_SET));
+		errnoBefore = errno;
+		TEST_ASSERT_EQUAL_INT(0, feof(f));
+		TEST_ASSERT_EQUAL(errnoBefore, errno);
+		pos += testdata_hugeSize / 50;
+		errno++;
+	}
+	TEST_ASSERT_EQUAL_INT(-1, fgetc(f));
+	errnoBefore = errno;
+	TEST_ASSERT_EQUAL_INT(1, feof(f));
+	TEST_ASSERT_EQUAL_INT(errnoBefore, errno);
+	fclose(f);
+	remove(filename);
+}
+
+
+TEST_GROUP_RUNNER(stdio_feof)
+{
+	RUN_TEST_CASE(stdio_feof, not_empty_all_modes);
+	RUN_TEST_CASE(stdio_feof, empty_all_modes);
+	RUN_TEST_CASE(stdio_feof, preserve_errno_huge_size)
+}
+
+
+TEST_GROUP(stdio_ftell);
+
+TEST_SETUP(stdio_ftell)
+{
+}
+
+
+TEST_TEAR_DOWN(stdio_ftell)
+{
+}
+
+
+TEST(stdio_ftell, correct_position_not_empty)
+{
+	FILE *f;
+	const char *filename = "test_stdio_ftell_not_empty";
+
+	for (int i = 0; i < sizeof(modes) / sizeof(modes[0]); ++i) {
+		char *data = testdata_createCharStr(ALL_ASCII);
+		f = fopen(filename, "w+");
+		TEST_ASSERT_NOT_NULL(f);
+		TEST_ASSERT_NOT_NULL(data);
+		TEST_ASSERT_NOT_EQUAL_INT_MESSAGE(EOF, fputs(data, f), tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, fclose(f), tellMode(modes[i]));
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_INT32_MESSAGE(-1, (int32_t)ftell(f), tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT32_MESSAGE((int32_t)ftell(f), (int32_t)ftello(f), tellMode(modes[i]));
+
+		f = fopen(filename, modes[i]);
+
+		TEST_ASSERT_NOT_NULL_MESSAGE(f, tellMode(modes[i]));
+		TEST_ASSERT_EQUAL_INT32_MESSAGE((int32_t)ftell(f), (int32_t)ftello(f), tellMode(modes[i]));
+
+		if (EQ("w", modes[i]) || EQ("wb", modes[i])) { /* can't read in "w" mode */
+			TEST_ASSERT_EQUAL_INT_MESSAGE(0, fclose(f), tellMode(modes[i]));
+			TEST_ASSERT_EQUAL_MESSAGE(0, remove(filename), tellMode(modes[i]));
+			free(data);
+			continue;
+		}
+
+		TEST_ASSERT_EQUAL_INT_MESSAGE(0, fseek(f, 0, SEEK_SET), tellMode(modes[i]));
+
+		{
+			int i = 1;
+			int errnoBefore;
+			char c;
+			while ((c = fgetc(f)) != -1) {
+				TEST_ASSERT_EQUAL_CHAR(c, data[i - 1]);
```

Comment:
What about this error, I see that it occurs on all targets except ia32, host-generic-pc. 
![image](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/116297248/a6c14220-5670-4008-918d-dab5811d896a)

Reply:
Seems not related to ftell nor fgetc function, removed the assertion that caused it.

Reply:
Even if it's not directly related to fread or fgetc, it would be worth to check whether it's something desired or not. If not, an issue should be reported.

As I see there are still problems with the last character of the extended ascii ?
![image](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/116297248/6769bdc3-ca59-4d40-9528-0d0171ec7660)